WasatchNET/AgilentSpectrometer.cs

```diff
@@ -0,0 +1,268 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Threading;
+using LibUsbDotNet;
+using LibUsbDotNet.Main;
+using SORS3CCA;
+
+namespace WasatchNET
+{
+    public class AgilentSpectrometer : Spectrometer
+    {
+        internal Wrapper wrapper;
+
+        internal AgilentSpectrometer(UsbRegistry usbReg) : base(usbReg)
+        {
+            wrapper = Wrapper.getInstance();
+            excitationWavelengthNM = 0;
+            triggerSource = TRIGGER_SOURCE.EXTERNAL;
+        }
+
+        public override uint scanAveraging
+        {
+            get { return (uint)wrapper.getScansToAverage(); }
+            set
+            {
+                lock (acquisitionLock)
+                {
+                    scanAveraging_ = value;
+                    wrapper.setScansToAverage((int)value);
+                }
+            }
+        }
+
+        override internal bool open()
+        {
+            wrapper.setLogEnabled(false);
+            eeprom = new EEPROM(this);
+
+            if (wrapper == null)
+            {
+                logger.info("Unable to instantiate SORS3CCA.Wrapper");
+                return false;
+            }
+
+            if (wrapper.init())
+            {
+                pixels = (uint)wrapper.getNumberOfPixels();
+
+                logger.info("found spectrometer with {0} pixels", pixels);
+
+                char[] ccaVer = new char[64];
+                wrapper.getCCAVersion(64, ref ccaVer);
+                string ccaVerStr = new string(ccaVer);
+                logger.info("CCA Version: {0}", ccaVerStr);
+
+                if (!eeprom.read())
+                {
+                    logger.error("Spectrometer: failed to GET_MODEL_CONFIG");
+                    wrapper.shutdown();
+                    return false;
+                }
+                logger.debug("back from reading EEPROM");
+
+                regenerateWavelengths();
+
+                return true;
+
+            }
+            else
+            {
+                logger.info("Unable to initialize SORS3CCA.Wrapper");
+                return false;
+            }
+
+        }
+
+        public override void close()
+        {
+            wrapper.shutdown();
+        }
+
+        public override double[] getSpectrum()
+        {
+            lock (acquisitionLock)
+            {
+                double[] specRaw = new double[pixels];
+
+                wrapper.getSpectrum((int)pixels, ref specRaw);
+
+                if (dark != null && dark.Length == specRaw.Length)
+                    for (int px = 0; px < pixels; px++)
+                        specRaw[px] -= dark_[px];
```

Comment:
It's not immediately obvious to me from a web review how dark differs from dark_, but if this compiles and works then okay.